starter-code/routes/auth.js

```diff
@@ -0,0 +1,69 @@
+const express = require('express');
+const bcrypt = require('bcrypt');
+const ensureLogin = require('connect-ensure-login');
+const passport = require('passport');
+
+const User = require('../models/user');
+
+const router = express.Router();
+const bcryptSalt = 10;
+
+// ENSURELOGIN
+// function authensure (req, res, next) {
+//   if (!req.user) {
+//     res.redirect('/login');
+//   } else {
+//     next();
+//   }
+// }
+
+//
+
+router.get('/private-page', ensureLogin.ensureLoggedIn(), (req, res) => {
+  res.render('passport/private', { user: req.user });
+});
+
+router.get('/signup', function (req, res, next) {
+  res.render('passport/signup');
+});
+
+router.post('/signup', (req, res, next) => {
+  const username = req.body.username;
+  const password = req.body.password;
+  const salt = bcrypt.genSaltSync(bcryptSalt);
+  const hashPass = bcrypt.hashSync(password, salt);
+
```

Comment:
You are missing important validation here.
First, no empty input:
```
if (username === "" || password === "") {
  res.render("passport/signup", { message: "Indicate username and password" });
  return;
}
```
Then, THE USER DOESN'T ALREADY EXISTS!!!:
```
User.findOne({ username }, "username", (err, user) => {
  if (user !== null) {
    res.render("passport/signup", { message: "The username already exists" });
    return;
  }

  var salt     = bcrypt.genSaltSync(bcryptSalt);
  var hashPass = bcrypt.hashSync(password, salt);

  var newUser = User({
    username,
    password: hashPass
  });

  newUser.save((err) => {
    if (err) {
      res.render("passport/signup", { message: "The username already exists" });
    } else {
      res.redirect("/login");
    }
  });
});
```
And you make the encryption when you are sure you are going to save the user. That way, you don't lose time if it isn't necessary.